base/setup/reactos/drivepage.c

```diff
@@ -48,28 +48,261 @@ static const INT  column_alignment[MAX_LIST_COLUMNS] = {LVCFMT_LEFT, LVCFMT_LEFT
 
 /* FUNCTIONS ****************************************************************/
 
-static INT_PTR CALLBACK
-MoreOptDlgProc(HWND hwndDlg,
-               UINT uMsg,
-               WPARAM wParam,
-               LPARAM lParam)
+/**
+ * @brief
+ * Sanitize a given string in-place, by
+ * removing any invalid character found in it.
+ **/
+static BOOL
+DoSanitizeText(
+    _Inout_ PWSTR pszSanitized)
+{
+    PWCHAR pch1, pch2;
+    BOOL bSanitized = FALSE;
+
+    for (pch1 = pch2 = pszSanitized; *pch1; ++pch1)
+    {
+        /* Skip any invalid character found */
+        if (!IS_VALID_INSTALL_PATH_CHAR(*pch1))
+        {
+            bSanitized = TRUE;
+            continue;
+        }
+
+        /* Copy over the valid ones */
+        *pch2 = *pch1;
+        ++pch2;
+    }
+    *pch2 = 0;
+
+    return bSanitized;
+}
+
+/**
+ * @brief
+ * Sanitize in-place any text found in the clipboard.
+ **/
+static BOOL
+DoSanitizeClipboard(
+    _In_ HWND hWnd)
+{
+    HGLOBAL hData;
+    LPWSTR pszText, pszSanitized;
+    BOOL bSanitized;
+
+    /* Protect read-only edit control from modification */
+    if (GetWindowLongPtrW(hWnd, GWL_STYLE) & ES_READONLY)
+        return FALSE;
+
+    if (!OpenClipboard(hWnd))
+        return FALSE;
+
+    hData = GetClipboardData(CF_UNICODETEXT);
+    pszText = GlobalLock(hData);
+    if (!pszText)
+    {
+        CloseClipboard();
+        return FALSE;
+    }
+
+    pszSanitized = _wcsdup(pszText);
+    GlobalUnlock(hData);
+    bSanitized = (pszSanitized && DoSanitizeText(pszSanitized));
+    if (bSanitized)
+    {
+        /* Update clipboard text */
+        DWORD cbData = (wcslen(pszSanitized) + 1) * sizeof(WCHAR);
+        hData = GlobalAlloc(GMEM_MOVEABLE | GMEM_SHARE, cbData);
+        pszText = GlobalLock(hData);
+        if (pszText)
+        {
+            CopyMemory(pszText, pszSanitized, cbData);
+            GlobalUnlock(hData);
+
+            SetClipboardData(CF_UNICODETEXT, hData);
+        }
+    }
+    free(pszSanitized);
+
+    CloseClipboard();
+    return bSanitized;
+}
+
+static VOID
+ShowErrorTip(
+    _In_ HWND hEdit)
+{
+    WCHAR szTitle[512];
+    WCHAR szText[512];
+
+    /* Load the resources */
+    LoadStringW(SetupData.hInstance, IDS_ERROR_INVALID_INSTALLDIR_CHAR_TITLE, szTitle, _countof(szTitle));
+    LoadStringW(SetupData.hInstance, IDS_ERROR_INVALID_INSTALLDIR_CHAR, szText, _countof(szText));
+
+    /* Show a warning balloon */
+    EDITBALLOONTIP balloon;
+    balloon.cbStruct = sizeof(balloon);
+    balloon.pszTitle = szTitle;
+    balloon.pszText  = szText;
+#if (_WIN32_WINNT < _WIN32_WINNT_VISTA)
+    balloon.ttiIcon  = TTI_ERROR;
+#else
+    balloon.ttiIcon  = 6; // TTI_ERROR_LARGE;
```

Comment:
assigning TTI_ERROR_LARGE insead of the hardcoded 6 is not possible? Because?

Reply:
Because we don't have the define in our headers yet.

Reply:
Which header do the reviewers think should that define be put to ultimately?

Reply:
commctrl.h, see commit 50986d8470c9930972183c323d7ef3425644cacc.